File: Programmers/[12952] N-Queen.py
def solution(N):
    global answer
    answer = 0
    line = [0] * N

    def check(i):
        for j in range(i):
            if line[i] == line[j] or abs(line[i] - line[j]) == (i - j):
                return False
        return True

    def place_queen(index):
        if index == N:
            global answer
            answer += 1
        else:
            for i in range(N):
                line[index] = i
                if check(index):
                    place_queen(index + 1)

    place_queen(0)
    return answer


if __name__ == "__main__":
    N = 12
    print(solution(N))

File: Programmers/[12952] N-Queen.java
package pg12952;

class Solution {

    static int N;
    static int[] array;
    static int answer;

    static boolean check(int count, int number) {
        for (int i = 0; i < count; i++) {
            if (array[i] == number) {
                return false;
            }
            if (Math.abs(array[i] - number) == Math.abs(i - count)) {
                return false;
            }
        }
        return true;
    }

    static void dfs(int count) {
        if (count == N) {
            answer += 1;
            return;
        }

        for (int i = 0; i < N; i++) {
            if (check(count, i)) {
                array[count] = i;
                dfs(count + 1);
            }
        }
    }

    public int solution(int n) {
        N = n;
        array = new int[N];

        dfs(0);

        return answer;
    }
}